[REVIEW] Add a utility function for producing an overflow-safe kernel launch grid configuration.

eyalroz opened the pull request:
Fixes issue #1394. Split off from PR #1366. Note: Depends on PR #1391 (specifically, the integer utilities code; not the bit manipulation stuff); so it can't be merged until PR #1391 was merged.

jrhemstad commented on cpp/src/utilities/cuda_utils.hpp:
```diff
@@ -0,0 +1,53 @@
+
+namespace cuda {
+
+inline constexpr auto form_naive_1d_grid(
```
The naming leaves something to be desired here. Also, Doxygen with a usage example would be nice.

eyalroz replied:
Do you have a specific idea regarding naming? How about "simple 1d grid"?

eyalroz replied:
I've redone this as a class instead of a function returning an anonymous struct. Also, renamed it. Tell me what you think.

jrhemstad commented on cpp/src/utilities/cuda_utils.hpp:
```diff
@@ -0,0 +1,72 @@
+ * @return An anonymous-type grid configuration, with a number of threads per block (the
+ * one specified
+ */
+class simple_1d_grid {
```
I'm thinking the naming could be even further simplified. 

What about just `1d_grid`?

Will there ever be a need for a `complex_1d_grid` that you'd need to differentiate by pre-pending `simple_`?

eyalroz replied:
Have you heard the `fresh fish sold here` story? I think [it](https://kathakids.com/folktales-funny-stories-hodja-fresh-fish-sold-here-367ad63de211) might apply here...

It's a simple grid; it's also a 1d grid. `simple_1d_grid` has a nice ring to it. Also, you could complicate a 1d grid formation, e.g. by reducing the block size if it's just a single block; or making the number of blocks a multiple of some value; etc.

harrism replied:
I think 1D grids are so simple that this class isn't needed at all.

harrism replied:
But if you must... please call it `grid_1d`.  If any specialized version comes up, it can be called `grid_1d_FEATUREX`. We don't need to call this one `grid_1d_simple`.

eyalroz replied:
I won't quibble over the name even though I like mine much better... so `grid_1d` it is.

eyalroz commented:
@harrism : If we need to specialize, we should probably use namespace rather than name suffixes/prefixes.

harrism commented:
> @harrism : If we need to specialize, we should probably use namespace rather than name suffixes/prefixes.

Hmm, `cudf::simple::grid_1` also seems unnecessary. :)

eyalroz commented:
@harrism : 
1. Renamed as per your suggestion (but called it a `grid_config` rather than a `grid`, since a grid may be construed to be what actually runs on the GPU.
2. Rephrased the Doxygen to reflect the fact that this is now a class, not a single function (so one comment for the class and another comment for the constructor).
3. Rebased on branch-0.7 for easier merging.

harrism commented:
Please don't force push to PR branches. It causes stuff to become unreachable. For example, if you now click the "view changes" button next to the review comments above, you get this:

<img width="1053" alt="image" src="https://user-images.githubusercontent.com/783069/56168369-59007b80-601e-11e9-87c1-dafb2737f02a.png">

harrism commented:
I like the name `grid_config_1d`!

eyalroz commented:
> Please don't force push to PR branches. It wreaks havoc with reviews. For example, if you now click the "view changes" button next to the review comments above, you get this:

But I need to force-push for replace irrelevant code, fixing last-minute bugs, rebasing etc...

harrism commented:
You "need" it? You can rebase and force push all you want until you create a PR. Once you have a PR with reviews, force pushing causes problems. Please use merge instead.

eyalroz commented:
@harrism : Ok, I'll respect the request and not force-push commits which have been reviewed. I still think it is easier to reconsider a force-pushed correction than to look at a sequence of commits and merges addressing reviews, bringing in changes on the target branch etc.
